orchagent/pfcwdorch.h

```diff
@@ -0,0 +1,121 @@
+#include "pfcactionhandler.h"
+
+extern "C" {
+#include "sai.h"
+}
+
+enum class PfcWdAction
+{
+    PFC_WD_ACTION_UNKNOWN,
+    PFC_WD_ACTION_FORWARD,
+    PFC_WD_ACTION_DROP,
+};
+
+template <typename DropHandler, typename ForwardHandler>
+class PfcWdOrch: public Orch
+{
+public:
+    PfcWdOrch(DBConnector *db, vector<string> &tableNames);
+    virtual ~PfcWdOrch(void);
+
+    virtual void doTask(Consumer& consumer);
+    virtual vector<sai_port_stat_t> getPortCounterIds(sai_object_id_t queueId);
+    virtual vector<sai_queue_stat_t> getQueueCounterIds(sai_object_id_t queueId);
+    virtual bool startWdOnPort(const Port& port,
+            uint32_t detectionTime, uint32_t restorationTime, PfcWdAction action) = 0;
+    virtual bool stopWdOnPort(const Port& port) = 0;
+
+    inline shared_ptr<ProducerStateTable> getPfcWdTable(void)
+    {
+        return m_pfcWdTable;
+    }
+
+    inline shared_ptr<Table> getCountersTable(void)
+    {
+        return m_countersTable;;
+    }
+
+private:
+    template <typename T>
+    static string counterIdsToStr(const vector<T> ids, string (*convert)(T));
+    static PfcWdAction deserializeAction(const string& key);
+    void createEntry(const string& key, const vector<FieldValueTuple>& data);
+    void deleteEntry(const string& name);
+    void registerInWdDb(const Port& port);
+    void unregisterFromWdDb(const Port& port);
+
+    shared_ptr<DBConnector> m_pfcWdDb = nullptr;
+    shared_ptr<DBConnector> m_countersDb = nullptr;
+    shared_ptr<ProducerStateTable> m_pfcWdTable = nullptr;
+    shared_ptr<Table> m_countersTable = nullptr;
+};
+
+template <typename DropHandler, typename ForwardHandler>
+class PfcWdSwOrch: public PfcWdOrch<DropHandler, ForwardHandler>
+{
+public:
+    PfcWdSwOrch(DBConnector *db, vector<string> &tableNames);
+    virtual ~PfcWdSwOrch(void);
+
+    virtual string getStormDetectionCriteria(void) = 0;
+
+    virtual bool startWdOnPort(const Port& port,
+            uint32_t detectionTime, uint32_t restorationTime, PfcWdAction action);
+    virtual bool stopWdOnPort(const Port& port);
+
+    //XXX Add port/queue state change event handlers
+private:
+    struct PfcWdQueueEntry
+    {
+        PfcWdQueueEntry(uint32_t detectionTime, uint32_t restorationTime,
+                PfcWdAction action, sai_object_id_t port, uint8_t idx);
+
+        const uint32_t c_detectionTime = 0;
+        const uint32_t c_restorationTime = 0;
+        const PfcWdAction c_action = PfcWdAction::PFC_WD_ACTION_UNKNOWN;
+
+        sai_object_id_t portId = SAI_NULL_OBJECT_ID;
+        // Remaining time till the next poll
+        uint32_t pollTimeLeft = 0;
+        uint8_t index = 0;
+        shared_ptr<PfcWdActionHandler> handler = { nullptr };
+    };
+
+    bool startWdOnQueue(sai_object_id_t queueId, uint8_t idx, sai_object_id_t portId,
+            uint32_t detectionTime, uint32_t restorationTime, PfcWdAction action);
+    bool stopWdOnQueue(sai_object_id_t queueId);
+    uint32_t getNearestPollTime(void);
```

Comment:
Do we have plan to change the polling interval? Or any idea how to meet the polling interval requirement based on current design?

Reply:
Not sure I understand. It's configurable via APP DB. There is a fixed polling time in syncd equal to 100 ms. We can change that if we need smaller polling time.

Reply:
I understand syncd has polling time equal to 100ms. But the current implementation of pfcwd doesn't query the database result from syncd every 100ms from my understanding.  On the contrary, the polling time for PFCWD depends on the detection time setting according to getNearestPollingTime implementation. Correct me if I am wrong.

Reply:
Let's consider the following example:
syncd has a hard coded polling period of 100 ms.
orchagent has a configured timer by a user - 200 ms for detection and 1000 ms for restoration.
In a regular flow, orchagent will poll queues every 200 ms.
In case of deadlock, orchagent will poll locked queues every 1000 ms, and it still has to poll regular ones every 200 ms. That's why we need to get nearest poll time so that it will be able to maintain all the timers.
Nearest poll time will never be less than 200 ms.
And having a hard coded minimal time in syncd was agreed during design with @lguohan .

Reply:
Hi Marian, I understand your point.  The logic is correct in general and I am fine with the syncd timer, though it will fail the requirement of granularity. We have the requirement to detect the storm within [detection_time, detect_time+poll_time] starting from when the storm happens. 

For instance, we have polling interval as 100ms, detection time as 200ms. Suppose a pfc storm start to happen at 50ms, we could detect it at 250ms. If we follow the current logic, we will detect it at 450ms. And we miss the requirement to detect it within [200, 300]

If you agree on this, we could leave it to phase 2.

Reply:
Lets leave it for P2. I'll prepare a more detailed description with examples for all cases that SW watchdog may be triggered and we can discuss it.

Reply:
BTW, if storm starts at 50 ms, we will detect it at 350 ms, so the actual range in software PFC WD is [detect time, detect_time * 2]